File: tabulation/javascript/howSum.js
function howSum(targetSum, numbers) {
    const table = Array(targetSum + 1).fill(null)
    table[0] = []

    for (let i = 0; i <= targetSum; i++) {
        if (table[i])
            for (let num of numbers)
                table[i + num] = [num, ...table[i]]
    }

    return table[targetSum]
}

console.log(howSum(7, [2, 3]))
console.log(howSum(51, [2, 4]))

File: tabulation/kotlin/howSum.kt
fun howSum(targetSum: Int, numbers: Array<Int>): List<Int>? {
    val table: Array<List<Int>?> = Array(targetSum + 1) { null }
    table[0] = listOf()

    for (i: Int in (0..targetSum)) {
        if (table[i] != null) {
            for (num: Int in numbers)
                if(i + num <= targetSum)
                    table[i + num] = table[i]?.plus(num)
        }
    }

    return table[targetSum]
}

fun main(): Unit {
    println(howSum(7, arrayOf(2, 3)))
    println(howSum(47, arrayOf(6, 2)))
}